### PR TITLE
IKASAN-1894 - Add ComponentFactory spec interface, ComponentFactory b…

### PR DESCRIPTION
**Pull request by kieronedwards**
…ase classes, and Jms Component Factory classes

### CODE REVIEW

File: ikasaneip/component-factory/spring/jms/src/main/java/org/ikasan/component/factory/jms/JmsComponentFactory.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+ *  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ *  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ *  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+ *  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *  ====================================================================
+ *
+ */
+
+package org.ikasan.component.factory.jms;
```
Comment (mitcje) [inline ikasaneip/component-factory/spring/jms/src/main/java/org/ikasan/component/factory/jms/JmsComponentFactory.java]: Should this not live in an AMQ specific package as it has AMQ specifics - its not just using JMS spec.

File: ikasaneip/component-factory/spring/jms/src/main/java/org/ikasan/component/factory/jms/consumer/JmsConsumerComponentFactory.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+ */
+public class JmsConsumerComponentFactory extends NonConfiguredResourceBaseComponentFactory<JmsContainerConsumer,
+        JmsConsumerConfiguration> {
+
+    @Resource
+    private JmsComponentFactory jmsComponentFactory;
+
+
+    @Override
+    public JmsContainerConsumer create(String nameSuffix, String configPrefix) {
```
Comment (mitcje) [inline ikasaneip/component-factory/spring/jms/src/main/java/org/ikasan/component/factory/jms/consumer/JmsConsumerComponentFactory.java]: This is spring specific - should this not live in a different package ? Where would we put the native JMS client for instance ?

File: ikasaneip/component-factory/spring/jms/src/main/java/org/ikasan/component/factory/jms/producer/JmsProducerComponentFactory.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ * @see org.ikasan.component.factory.jms.consumer.JmsConsumerComponentFactory
+ */
+public class JmsProducerComponentFactory extends NonConfiguredResourceBaseComponentFactory<JmsTemplateProducer,
+        JmsProducerConfiguration> {
+
+    @Resource
+    private JmsComponentFactory jmsComponentFactory;
+
+    @Override
+    public JmsTemplateProducer create(String nameSuffix, String configPrefix) {
```
Comment (mitcje) [inline ikasaneip/component-factory/spring/jms/src/main/java/org/ikasan/component/factory/jms/producer/JmsProducerComponentFactory.java]: Again spring specific - where would we implement the native JMS publisher version?

**Review by mitcje — CHANGES_REQUESTED** [review]
@kieronedwards  please can you clarify the comments below - should we not be isolating spring specific JMS facets as we do have a native JMS client as well.

File: ikasaneip/component-factory/spring/jms/src/main/java/org/ikasan/component/factory/jms/JmsComponentFactory.java (hunk 1 above)
Comment (kieronedwards) [reply ikasaneip/component-factory/spring/jms/src/main/java/org/ikasan/component/factory/jms/JmsComponentFactory.java]: @mitcje  This is has proven to work with both AMQ and JBOSS Messaging we can extend to others as needed.

File: ikasaneip/component-factory/spring/jms/src/main/java/org/ikasan/component/factory/jms/consumer/JmsConsumerComponentFactory.java (hunk 2 above)
Comment (kieronedwards) [reply ikasaneip/component-factory/spring/jms/src/main/java/org/ikasan/component/factory/jms/consumer/JmsConsumerComponentFactory.java]: correct - all component factory implementations live under the component-factory spring maven module. We could rename the lib to include spring in its name - if that works ? 

As an aside I did a quick recce of builder its has multiple spring dependencies even in the non spring module (for example the harvester pulls in spring-core).

File: ikasaneip/component-factory/spring/jms/src/main/java/org/ikasan/component/factory/jms/producer/JmsProducerComponentFactory.java (hunk 3 above)
Comment (kieronedwards) [reply ikasaneip/component-factory/spring/jms/src/main/java/org/ikasan/component/factory/jms/producer/JmsProducerComponentFactory.java]: @mitcje see above comment

**Review by mitcje — APPROVED** [review]

**Pull request merged**